component/src/main/java/org/wso2/extension/siddhi/io/ibmmq/source/IBMMQConnectionRetryHandler.java

```diff
@@ -0,0 +1,122 @@
+import javax.jms.JMSException;
+
+/**
+ * This class tries to connect to MQ provider until the maximum re-try count meets.
+ */
+class IBMMQConnectionRetryHandler {
+    /**
+     * This {@link IBMMessageConsumerThread} instance represents the MQ receiver that asked for retry.
+     */
+    private IBMMessageConsumerThread messageConsumer;
+
+    private static final Logger logger = LoggerFactory.getLogger(IBMMQConnectionRetryHandler.class);
+
+    /**
+     * Current retry interval in milliseconds.
+     */
+    private long currentRetryInterval;
+
+    /**
+     * Initial retry interval in milliseconds.
+     */
+    private long retryInterval;
+
+    /**
+     * Current retry count.
+     */
+    private int retryCount = 0;
+
+    /**
+     * Maximum retry count.
+     */
+    private int maxRetryCount;
+
+    /**
+     * States whether a isOnRetrying is in progress.
+     */
+    private volatile boolean isOnRetrying = false;
+
+    /**
+     * Creates a IBMMQ connection retry handler.
+     *
+     * @param messageConsumer IBMMQ message consumer that needs to retry.
+     * @param retryInterval   Retry interval between.
+     * @param maxRetryCount   Maximum retries.
+     */
+    IBMMQConnectionRetryHandler(IBMMessageConsumerThread messageConsumer, long retryInterval, int maxRetryCount) {
+        this.messageConsumer = messageConsumer;
+        this.retryInterval = retryInterval;
+        this.maxRetryCount = maxRetryCount;
+        this.currentRetryInterval = retryInterval;
+    }
+
+    /**
+     * To retry the isOnRetrying to connect to MQ provider.
+     *
+     * @return True if isOnRetrying was successful.
+     * @throws JMSException MQ Connector Exception.
+     */
+    boolean retry() {
+        if (isOnRetrying) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Retrying is in progress from a different thread, hence not isOnRetrying");
+            }
+            return false;
+        } else {
+            isOnRetrying = true;
+        }
+
+        while (true) {
+            try {
+                retryCount++;
+                messageConsumer.shutdownConsumer();
+                messageConsumer.connect();
+                logger.info("Connected to the message broker to " + messageConsumer.getQueueName()
+                        + "after isOnRetrying for " + retryCount + " time(s)");
+                retryCount = 0;
+                currentRetryInterval = retryInterval;
+                isOnRetrying = false;
+                return true;
+            } catch (JMSException e) {
+                messageConsumer.shutdownConsumer();
+                logger.error("Retry connection attempt " + retryCount + " to MQ Provider failed. Retry will be " +
+                        "attempted again after " +
+                        TimeUnit.SECONDS.convert(currentRetryInterval, TimeUnit.MILLISECONDS) + " seconds");
+                try {
+                    TimeUnit.MILLISECONDS.sleep(currentRetryInterval);
```

Comment:
Shall we use proper retry logic as in https://github.com/siddhi-io/siddhi/blob/master/modules/siddhi-core/src/main/java/io/siddhi/core/table/Table.java#L385?

Reply:
fixed with 27879ffaa71f98759d5d0f4163e8a40e0c7812a1